src/envs/repeated_nextto_painting.py

```diff
@@ -0,0 +1,182 @@
+        target_obj = self._get_object_at_xyz(state, x, y, z)
+        # if PaintingEnv pick was sucessful check if the object was next to
+        if target_obj is not None and state.get(
+                target_obj, "held") == 0.0 and next_state.get(
+                    target_obj, "held") == 1.0:
+            # Added cannot pick if obj is too far
+            if abs(state.get(self._robot, "y") \
+                - state.get(target_obj, "pose_y")) \
+                >= self.nextto_thresh:
+                return state
+        return next_state
+
+    def _transition_nextto_place(self, state: State, action: Action) -> State:
+        # Action args are target pose for held obj
+        _, y, _ = action.arr[:3]
+        next_state = self._transition_place(state, action)
+        #  Added restriction on place if not close enough to location
+        if abs(state.get(self._robot, "y") - y) \
+            >= self.nextto_thresh:
+            return state
+        return next_state
+
+    def _transition_move(self, state: State, action: Action) -> State:
+        # Action args are target y for robot
+        y = action.arr[1]
+        next_state = state.copy()
+        # Execute move
+        next_state.set(self._robot, "y", y)
+        held_obj = self._get_held_object(state)
+        if held_obj is not None:
+            next_state.set(held_obj, "pose_y", y)
+        return next_state
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {
+            self._InBox, self._InShelf, self._IsBoxColor, self._IsShelfColor,
+            self._GripperOpen, self._OnTable, self._HoldingTop,
+            self._HoldingSide, self._Holding, self._IsWet, self._IsDry,
+            self._IsDirty, self._IsClean, self._NextTo, self._NextToBox,
+            self._NextToShelf, self._NextToNothing
+        }
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {
+            self._Pick, self._Wash, self._Dry, self._Paint, self._Place,
+            self._OpenLid, self._MoveToObj, self._MoveToBox, self._MoveToShelf
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # Actions are 8-dimensional vectors:
+        # [x, y, z, grasp, pickplace, water level, heat level, color]
+        # Note that pickplace is 1 for pick, -1 for place, and 0 otherwise,
+        # while grasp, water level, heat level, and color are in [0, 1].
+        # Changed lower bound for z to 0.0 for RepeatedNextToPainting
+        # This is needed to check affinity of the move action
+        lowers = np.array(
+            [self.obj_x - 1e-2, self.env_lb, 0.0, 0.0, -1.0, 0.0, 0.0, 0.0],
+            dtype=np.float32)
+        uppers = np.array([
+            self.obj_x + 1e-2, self.env_ub, self.obj_z + 1e-2, 1.0, 1.0, 1.0,
+            1.0, 1.0
+        ],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def _OnTable_holds(self, state: State, objects: Sequence[Object]) -> bool:
```

Comment:
I can see how this definition is more correct in some sense, but does something break if we use the original definition?

Reply:
Since, in the painting environment the held objects remained on the table. I had to add a change in z (to simulate lifting the object), because in the repeated_nextto_painting environment there are move options and if you move to another block the locations of the blocks would be the same. Nothing really breaks, I believe I could've defined the nsrts with the object being On_Table the whole time, but I didn't like objects having the same location after movement.